src/server.rs

```diff
@@ -0,0 +1,189 @@
+use crate::config::Config;
+use bytes::{BufMut, BytesMut};
+use std::{borrow::Cow, fmt::Debug, str};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream, ToSocketAddrs},
+};
+use std::fmt::Formatter;
+use std::{
+    io::{BufRead, Cursor, Seek, SeekFrom},
+    str::FromStr,
+};
+use tokio::{
+    io::{AsyncBufReadExt, Error, ErrorKind},
+    stream::StreamExt,
+};
+use tokio_util::codec::{Decoder, Encoder, FramedRead};
+use std::path::PathBuf;
+use crate::networking::Command;
+
+#[derive(Debug)]
+pub enum ClientCommand {
+    Handshake(String),          // HI#<hdid:String>#%
+    ClientVersion(u32, String, String),      // ID#<pv:u32>#<software:String>#<version:String>#%
+    KeepAlive,          // CH
+    AskListLengths,     // askchaa
+    AskListCharacters,  // askchar
+    CharacterList(u32),      // AN#<page:u32>#%
+    EvidenceList(u32),       // AE#<page:u32>#%
+    MusicList(u32),          // AM#<page:u32>#%
+    AO2CharacterList,   // AC#%
+    AO2MusicList,       // AM#%
+    AO2Ready,           // RD#%
+    SelectCharacter(u32, u32, String),    // CC<client_id:u32>#<char_id:u32#<hdid:String>#%
+    ICMessage,          // MS
+    OOCMessage(String, String),         // CT#<name:String>#<message:String>#%
+    PlaySong(u32, u32),           // MC#<song_name:u32>#<???:u32>#%
+    WTCEButtons(String),        // RT#<type:String>#%
+    SetCasePreferences(String, CasePreferences), // SETCASE#<cases:String>#<will_cm:boolean>#<will_def:boolean>#<will_pro:boolean>#<will_judge:boolean>#<will_jury:boolean>#<will_steno:boolean>#%
+    CaseAnnounce(String, CasePreferences),       // CASEA
+    Penalties(u32, u32),          // HP#<type:u32>#<new_value:u32>#%
+    AddEvidence(EvidenceArgs),        // PE#<name:String>#<description:String>#<image:String>#%
+    DeleteEvidence(u32),     // DE#<id:u32>#%
+    EditEvidence(u32, EvidenceArgs),       // EE#<id:u32>#<name:String>#<description:String>#<image:String>#%
+    CallModButton(Option<String>),      // ZZ?#<reason:String>?#%
+}
+
+impl Command for ClientCommand {
+    fn from_protocol(name: String, mut args: Vec<String>) -> Result<Self, anyhow::Error> {
+        let args_len = args.len();
+        match name.as_str() {
+            "HI" => {
+                if args_len != 1 {
+                    anyhow::bail!("Amount of arguments for command HANDSHAKE does not match!")
+                }
+
+                Ok(Self::Handshake(args.remove(0)))
+            },
+            _ => Ok(Self::ICMessage)
```

Comment:
I think it's better to use only match. Something like this:
```suggestion
        match (name.as_str(), args.as_mut_slice()) {
            ("HI", [arg]) => Ok(Self::Handshake(mem::take(arg)),
            (cmd, _) => anyhow::bail!("Amount of arguments for command {} does not match!", cmd),
```